src/Compilers/CSharp/Test/Symbol/Symbols/ExtendedPartialMethodsTests.cs

```diff
@@ -0,0 +1,1794 @@
+            comp.VerifyDiagnostics(
+                // (4,34): error CS8793: Partial method 'Program.M(ref object)' must have an implementation part because it has accessibility modifiers.
+                //     internal static partial void M(ref object o);
+                Diagnostic(ErrorCode.ERR_PartialMethodWithAccessibilityModsMustHaveImplementation, "M").WithArguments("Program.M(ref object)").WithLocation(4, 34),
+                // (5,34): error CS0663: 'Program' cannot define an overloaded method that differs only on parameter modifiers 'out' and 'ref'
+                //     internal static partial void M(out object o);
+                Diagnostic(ErrorCode.ERR_OverloadRefKind, "M").WithArguments("Program", "method", "out", "ref").WithLocation(5, 34));
+        }
+
+        [Fact]
+        public void ConflictingOverloads_RefOut_02()
+        {
+            const string text = @"
+partial class Program
+{
+    internal static partial void M(ref object o);
+    internal static partial void M(ref object o) { }
+    internal static partial void M(out object o);
+    internal static partial void M(out object o) { o = null; }
+}
+";
+            var comp = CreateCompilation(text, parseOptions: TestOptions.RegularWithExtendedPartialMethods);
+            comp.VerifyDiagnostics(
+                // (6,34): error CS0663: 'Program' cannot define an overloaded method that differs only on parameter modifiers 'out' and 'ref'
+                //     internal static partial void M(out object o);
+                Diagnostic(ErrorCode.ERR_OverloadRefKind, "M").WithArguments("Program", "method", "out", "ref").WithLocation(6, 34));
+        }
+
+        [Fact]
+        public void RefReturn()
+        {
+            const string text = @"
+using System;
+
+partial class Program
+{
+    static int i = 1;
+    private static partial ref int M();
+    private static partial ref int M() => ref i;
+
+    static void Main()
+    {
+        ref int local = ref M();
+        Console.Write(local);
+        i++;
+        Console.Write(local);
+    }
+}
+";
+            var verifier = CompileAndVerify(
+                text,
+                parseOptions: TestOptions.RegularWithExtendedPartialMethods,
+                expectedOutput: "12");
+            verifier.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ProtectedAccessibility()
+        {
+            const string text1 = @"
+using System;
+
+partial class Base
+{
+    protected static partial void M1();
+    protected static partial void M1() { Console.Write(1); }
+}";
+
+            const string text2 = @"
+class Derived : Base
+{
+    static void Main()
+    {
+        M1();
+    }
+}";
+            var verifier = CompileAndVerify(
+                new[] { text1, text2 },
+                parseOptions: TestOptions.RegularWithExtendedPartialMethods,
+                expectedOutput: "1");
+            verifier.VerifyDiagnostics();
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider testing access outside the type hierarchy as well.

Reply:
Added more accessibility tests.